RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc

```diff
@@ -0,0 +1,253 @@
+}
+
+PixelClusterTagInfoProducer::~PixelClusterTagInfoProducer() {}
+
+// ------------ method called to produce the data  ------------
+void PixelClusterTagInfoProducer::produce(edm::StreamID iID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  // Declare produced collection
+  auto pixelTagInfo = std::make_unique<reco::PixelClusterTagInfoCollection>();
+
+  // Open jet collection
+  edm::Handle<edm::View<reco::Jet> > collectionJets;
+  iEvent.getByToken(m_jets, collectionJets);
+
+  // Count jets above threshold
+  int nJets(0);
+  for (auto jetIt = collectionJets->begin(); jetIt != collectionJets->end(); ++jetIt) {
+    if (jetIt->pt() > m_minJetPt)
+      nJets++;
+  }
+
+  // Get primary vertex in the event
+  edm::Handle<reco::VertexCollection> collectionPVs;
+  iEvent.getByToken(m_vertices, collectionPVs);
+  reco::VertexCollection::const_iterator firstPV = collectionPVs->begin();
+
+  // If no suitable Jet and PV is available, skip the event without opening pixel collection
+  if (collectionPVs->empty() || nJets <= 0) {
+    iEvent.put(std::move(pixelTagInfo));
+    return;
+  }
+
+  // Open Pixel Cluster collection
+  edm::Handle<edmNew::DetSetVector<SiPixelCluster> > collectionHandle;
+  iEvent.getByToken(m_pixelhit, collectionHandle);
+  const edmNew::DetSetVector<SiPixelCluster>& collectionClusters(*collectionHandle);
+
+  // Open Geometry
+  edm::ESHandle<TrackerGeometry> geom;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geom);
+  const TrackerGeometry& theTracker(*geom);
+
+  // Retrieve tracker topology from geometry
+  edm::ESHandle<TrackerTopology> tTopoH;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoH);
+  const TrackerTopology* tTopo = tTopoH.product();
+
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (auto const& detUnit : collectionClusters) {
+    if (detUnit.empty())
+      continue;
+    DetId detId = DetId(detUnit.detId());  // Get the Detid object
+    unsigned int detType = detId.det();    // det type, pixel = 1
+    if (detType != 1)
+      continue;                             // Consider only pixels
+    unsigned int subid = detId.subdetId();  // Subdetector type, pix barrel = 1, forward = 2
+    if (!(subid == PixelSubdetector::PixelBarrel || (m_addFPIX && subid == PixelSubdetector::PixelEndcap)))
+      continue;
+    unsigned int layer = tTopo->layer(detId);  // The layer index is in range 1-4 or 1-3
+    if (layer == 0 || layer > m_nLayers)
+      continue;
+
+    // Get the geom-detector
+    const auto* geomDet = theTracker.idToDet(detId);
+    const auto* topol = &geomDet->topology();
+
+    for (auto const& clUnit : detUnit) {
+      // get global position of the cluster
+      LocalPoint lp = topol->localPosition(MeasurementPoint(clUnit.x(), clUnit.y()));
+      GlobalPoint clustgp = geomDet->surface().toGlobal(lp);
+      if (m_minADC > 0 and clUnit.charge() < m_minADC)
```

Comment:
Please move at the beginning of the loop: no reason to compute ```lp``` and ```clustgp``` if below the adc threshold